server/src/main/java/org/opensearch/index/query/BoolQueryBuilder.java

```diff
@@ -460,4 +472,85 @@ public void visit(QueryBuilderVisitor visitor) {
+        // every doc has exactly 1 value for this field.
+        List<LeafReaderContext> leafReaderContexts = getLeafReaderContexts(queryRewriteContext);
+        if (leafReaderContexts == null || leafReaderContexts.isEmpty()) {
+            return false;
+        }
+
+        boolean changed = false;
+        // For now, only handle the case where there's exactly 1 range query for this field.
+        Map<String, Integer> fieldCounts = new HashMap<>();
+        Set<RangeQueryBuilder> rangeQueries = new HashSet<>();
+        for (QueryBuilder clause : mustNotClauses) {
+            if (clause instanceof RangeQueryBuilder rq) {
+                fieldCounts.merge(rq.fieldName(), 1, Integer::sum);
+                rangeQueries.add(rq);
+            }
+        }
+
+        for (RangeQueryBuilder rq : rangeQueries) {
+            String fieldName = rq.fieldName();
+            if (fieldCounts.getOrDefault(fieldName, 0) == 1) {
+                // Check that all docs on this field have exactly 1 value, otherwise we can't perform this rewrite
+                if (checkAllDocsHaveOneValue(leafReaderContexts, fieldName)) {
+                    List<RangeQueryBuilder> complement = rq.getComplement();
+                    if (complement != null) {
+                        BoolQueryBuilder nestedBoolQuery = new BoolQueryBuilder();
+                        nestedBoolQuery.minimumShouldMatch(1);
+                        for (RangeQueryBuilder complementComponent : complement) {
+                            nestedBoolQuery.should(complementComponent);
+                        }
+                        newBuilder.must(nestedBoolQuery);
+                        newBuilder.mustNotClauses.remove(rq);
+                        changed = true;
+                    }
+                }
+            }
+        }
+
+        if (minimumShouldMatch == null && changed) {
+            if ((!shouldClauses.isEmpty()) && mustClauses.isEmpty() && filterClauses.isEmpty()) {
+                // If there were originally should clauses and no must/filter clauses, null minimumShouldMatch is set to a default of 1
+                // within Lucene.
+                // But if there was originally a must or filter clause, the default is 0.
+                // If we added a must clause due to this rewrite, we should respect what the original default would have been.
+                newBuilder.minimumShouldMatch(1);
+            }
+        }
+        return changed;
+    }
+
+    private List<LeafReaderContext> getLeafReaderContexts(QueryRewriteContext queryRewriteContext) {
+        if (queryRewriteContext == null) return null;
+        QueryShardContext shardContext = queryRewriteContext.convertToShardContext();
+        if (shardContext == null) return null;
+        IndexSearcher indexSearcher = shardContext.searcher();
+        if ((indexSearcher instanceof ContextIndexSearcher cis)) {
+            return cis.getLeafContexts();
+        }
```

Comment:
You don't need to check for `ContextIndexSearcher` here. You can safely return `indexSearcher.getIndexReader().leaves()`.